### ISSUE TITLE
Make dependencies for uix/memo and uix/callback Clojurescript friendly

### ISSUE THREAD

**Issue by caleb** [roman01la/uix#21]
Right now `uix/memo` and `uix/callback` just directly call React's underlying `useMemo` and `useCallback` which will trigger a re-calculation of the value/callback on ever render when used with Clojurescript datastructures.

What do you think about re-working these to use the same helpers as `uix/effect!` and `uix/layout-effect!`.

Here is something that I found works in my project to prevent unnecessary re-calculations:

```clojure
(defn callback
  "Takes function f and optional vector of dependencies, and returns f."
  ([f]
   (callback f nil))
  ([f deps]
    #?(:cljs (with-deps-check
               [prev-deps*]
               (react/useCallback (fn [& args]
                                    (reset! prev-deps* deps)
                                    (apply f args))
                                  (maybe-js-deps @prev-deps*))
               deps)
       :clj  f)))

(defn memo
  "Takes function f and optional vector of dependencies, and returns memoized f."
  ([f]
   (memo f nil))
  ([f deps]
    #?(:cljs (with-deps-check
               [prev-deps*]
               (react/useMemo (fn []
                                (reset! prev-deps* deps)
                                (f))
                              (maybe-js-deps @prev-deps*))
               deps)
       :clj  (f))))
```

(The above version of `memo` also calls `f` in the `clj` branch, since `useMemo` returns the value from the function, not the function itself, right now `uix/memo` returns the function itself in `memo` when run in Clojure)

**Comment by roman01la** [comment]
Yes. This is definitely needed.

One thing I'd want to have is tests that verifies that deps check actually works. This should be tests in the browser.

**Comment by caleb** [comment]
Good idea. Browser testing with React and Clojurescript is completely new to me, so I'll have to learn how to set that up.

**Comment by roman01la** [comment]
fixed in 057c6f966
